The active Launch Target is now saved in the ICBuildConfiguration

betamaxbandit opened the pull request:
In addition to the active launch mode and toolchain, the active Launch Target is now used to determine the Core Build Configuration (ICBuildConfiguration).

The build output folder name now ends with the name of the Launch Target (eg: cmake.debug.win32.x86_64.Local)

Addresses Issue: CDT CMake Improvements #1000, IDE-82683-REQ-016 Launch Targets used in ICBuildConfiguration

jonahgraham commented on core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/CBuildConfiguration.java:
```diff
@@ -208,6 +252,105 @@ public String getLaunchMode() {
+	 */
+	protected LaunchTargetInfo getLaunchTargetInfo(String launchTargetId) {
+		if (launchTargetId != null) {
+			String[] split = launchTargetId.split(LAUNCH_TARGET_ID_SEPARATOR);
```
Are IDs guaranteed to never have this separator by design? Instead of the joining and splitting, can you save the typeId to one field and the Id to another and then there is never ambiguity in serialization of parsing.

jonahgraham reviewed (commented):
I have done a first pass review. This looks reasonable.

In the call we discussed `null` vs `NULL_TARGET` vs `Local` target all fallback. I think it should `Local` as the fallback everywhere that is possible.

For example

```diff
diff --git a/core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/CBuildConfiguration.java b/core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/CBuildConfiguration.java
index e9c51160a3..c8be536350 100644
--- a/core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/CBuildConfiguration.java
+++ b/core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/CBuildConfiguration.java
@@ -271,6 +271,12 @@ public abstract class CBuildConfiguration extends PlatformObject implements ICBu
 				return target;
 			}
 		}
+
+		ILaunchTarget target = launchTargetManager.getLaunchTarget(ILaunchTargetManager.localLaunchTargetTypeId,
+				ILaunchTargetManager.localLaunchTargetId);
+		if (target != null) {
+			return target;
+		}
 		throw new CoreException(
 				Status.error(String.format("Error, Launch Target '%s' not found", launchTargetId), null)); //$NON-NLS-1$
 	}
diff --git a/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/Messages.java b/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/Messages.java
index 3e0163cb8f..df089b6903 100644
--- a/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/Messages.java
+++ b/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/Messages.java
@@ -21,7 +21,6 @@ public class Messages extends NLS {
 	public static String LaunchBarManager_0;
 	public static String LaunchBarManager_1;
 	public static String LaunchBarManager_2;
-	public static String LocalTarget_name;
 	public static String OK;
 
 	static {
diff --git a/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/messages.properties b/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/messages.properties
index 52f961bd86..3a056dbce4 100644
--- a/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/messages.properties
+++ b/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/messages.properties
@@ -15,5 +15,4 @@ ILaunchTarget_notSupported=getWorkingCopy is not supported for NULL_TARGET
 LaunchBarManager_0=Launch Bar Initialization
 LaunchBarManager_1=Active descriptor must be in the map of descriptors
 LaunchBarManager_2=Mode is not supported by descriptor
-LocalTarget_name=Local
 OK=OK
diff --git a/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/target/LocalLaunchTargetProvider.java b/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/target/LocalLaunchTargetProvider.java
index b010d43881..0c8140f822 100644
--- a/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/target/LocalLaunchTargetProvider.java
+++ b/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/internal/target/LocalLaunchTargetProvider.java
@@ -11,7 +11,6 @@
 package org.eclipse.launchbar.core.internal.target;
 
 import org.eclipse.core.runtime.Platform;
-import org.eclipse.launchbar.core.internal.Messages;
 import org.eclipse.launchbar.core.target.ILaunchTarget;
 import org.eclipse.launchbar.core.target.ILaunchTargetManager;
 import org.eclipse.launchbar.core.target.ILaunchTargetProvider;
@@ -23,9 +22,9 @@ public class LocalLaunchTargetProvider implements ILaunchTargetProvider {
 	@Override
 	public void init(ILaunchTargetManager targetManager) {
 		if (targetManager.getLaunchTarget(ILaunchTargetManager.localLaunchTargetTypeId,
-				Messages.LocalTarget_name) == null) {
+				ILaunchTargetManager.localLaunchTargetId) == null) {
 			ILaunchTarget target = targetManager.addLaunchTarget(ILaunchTargetManager.localLaunchTargetTypeId,
-					Messages.LocalTarget_name);
+					ILaunchTargetManager.localLaunchTargetId);
 			ILaunchTargetWorkingCopy wc = target.getWorkingCopy();
 			wc.setAttribute(ILaunchTarget.ATTR_OS, Platform.getOS());
 			wc.setAttribute(ILaunchTarget.ATTR_ARCH, Platform.getOSArch());
diff --git a/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/target/ILaunchTargetManager.java b/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/target/ILaunchTargetManager.java
index ebb83bef2e..c16094442c 100644
--- a/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/target/ILaunchTargetManager.java
+++ b/launchbar/org.eclipse.launchbar.core/src/org/eclipse/launchbar/core/target/ILaunchTargetManager.java
@@ -26,6 +26,7 @@ public interface ILaunchTargetManager {
 	 * represents that machine.
 	 */
 	final String localLaunchTargetTypeId = "org.eclipse.launchbar.core.launchTargetType.local"; //$NON-NLS-1$
+	final String localLaunchTargetId = "Local"; //$NON-NLS-1$
 
 	/**
 	 * The list of all launch targets.
```

You could add a new method to ILaunchTargetManager that always returns the Local launch.

Note: in the patch above I removed "Local" from the messages. This string must not be translated because it is being used as an ID. That means it must be constant even with different language files. (Back in time "Local" was the *name* but that changed in 7a1e27eb12b045e091e9e3eca053818129d8fadf) A proper label provider for local can be added with the org.eclipse.launchbar.ui.launchTargetTypeUI extension point.

betamaxbandit commented:
> null vs NULL_TARGET vs Local,
I've following most of your suggestion in this comment.

1) I've added ILaunchTargetManager.getLocalLaunchTarget()
2) Replaced all my uses of ILaunchTarget.NULL_TARGET with 1) above.
3) Removed the LocalTarget_name from messages and added the new localLaunchTargetId const.
4) However for the item containing the following block of code, I do not understand the idea:
+
+		ILaunchTarget target = launchTargetManager.getLaunchTarget(ILaunchTargetManager.localLaunchTargetTypeId,
+				ILaunchTargetManager.localLaunchTargetId);
+		if (target != null) {
+			return target;
+		}

Are you suggesting that in CBuildConfiguration.getLaunchTargetFromProperties() if the LT is not found in properties, then provide the Local LT as a fallback and thereby remove the coreexception throw? I'm not sure about that, as using a default in this case may  hide any erroneous operation.

jonahgraham commented:
>  I'm not sure about that, as using a default in this case may hide any erroneous operation.

OK - that place seems unsuitable. I was trying to code up what getting the Local looked like.

betamaxbandit commented:
Hi @jonahgraham ,
I've done many of the items we discussed in our call today. However, I ran out of time today and still have a few things remaining*.
Please could you review changes as they stand?



*:TODO:
- Complete ## API Changes in CDT 12.0. section of CHANGELOG-API.md
- Design an api to get the ICBuildConfiguration name, generated in ICBuildConfigurationProvider.createBuildConfiguration, which is duplicated in the following classes, almost identically:
  CMakeBuildConfigurationProvider
  AutotoolsBuildConfigurationProvider
  MakefileBuildConfigurationProvider
  MesonBuildConfigurationProvider
- Complete CBuildConfigurationManagerTests

jonahgraham commented on core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/CBuildConfiguration.java:
```diff
@@ -208,6 +259,42 @@ public String getLaunchMode() {
+		if (target != null) {
+			return target;
+		}
+		throw new CoreException(Status.error("Error, Launch Target not found", null)); //$NON-NLS-1$
```
I've had a bit more time to play with this. Throwing an exception here like this is very problematic as it generates endless exceptions in the log if thrown from somewhere that catches and logs it. But worse, if the exception is thrown from a Job such as the call to `configManager.getBuildConfiguration` in `CoreBuildLaunchBarTracker.setActiveBuildConfig(...).new Job() {...}.run(IProgressMonitor)` we get endless pop-ups too.

The real use case that you should be able to test/simulate is you launch a workspace and a launch target type id you previously had available no longer available. It also affects if the type id still exists, but the launch target id is no longer available. The latter happens when docker container stops being available.

Steps:

1. Create a CMake project and build/etc it so that it ends up in `.metadata/.plugins/org.eclipse.core.runtime/.settings/org.eclipse.cdt.core.prefs`
2. Quit Eclipse
3. Edit `.metadata/.plugins/org.eclipse.core.runtime/.settings/org.eclipse.cdt.core.prefs` to make the `launchTarget.typeId` invalid (simulating it being uninstalled or similar)
4. Start Eclipse again and note the log has a few errors in it
5. Try to do things such as:
* Change the launch config dropdown in the launch bar
* Create a new manual launch configuration the traditional Launch Configuration GUI
* Maybe others as I have seen the pop-up come in other places too

This is the pop-up I get:

![image](https://github.com/user-attachments/assets/9033c6f0-db6a-49e3-adc5-a682cb72c441)

This leaves me with no way to really resolve the problem. If the pop-up was one time **and** it explained the problem a bit that would be ok. But repeatedly without explaining is problematic.

What you can do is this (key part is STATUS_BUILD_CONFIG_NOT_VALID):

```suggestion
		throw new CoreException(new Status(IStatus.ERROR, CCorePlugin.PLUGIN_ID,
				CCorePlugin.STATUS_BUILD_CONFIG_NOT_VALID, "Error, Launch Target not found", null)); //$NON-NLS-1$
```

which will add the configuration to CBuildConfigurationManager.noConfigs and suppress the error entirely. But when something tries to use it later the dreaded `Build not configured correctly` error appears because it is noConfigs.

betamaxbandit replied:
Done.
I've used your STATUS_BUILD_CONFIG_NOT_VALID suggestion for launch mode and launch target now.

I've run your test steps and confirm I don't get a pop up; I do get the expected Build not configured correctly.

jonahgraham reviewed (commented):
@betamaxbandit I have done a partial review. But I ran out of time to complete it and ran into a hard error (the big comment about throwing exceptions).

betamaxbandit commented:
TODO:

Complete ## API Changes in CDT 12.0. section of CHANGELOG-API.md
Design an api to get the ICBuildConfiguration name, generated in ICBuildConfigurationProvider.createBuildConfiguration, which is duplicated in the following classes, almost identically:
CMakeBuildConfigurationProvider
AutotoolsBuildConfigurationProvider
MakefileBuildConfigurationProvider
MesonBuildConfigurationProvider
Complete CBuildConfigurationManagerTests

jonahgraham commented on core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/CCorePlugin.java:
```diff
@@ -125,7 +125,7 @@ public class CCorePlugin extends Plugin {
 
 	/**
-	 * Toolchain not found when inflating a build configuration.
+	 * Toolchain, launch mode or launch target not found when inflating a build configuration.
```
:+1: for making this update too.

jonahgraham commented on core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/CBuildConfiguration.java:
```diff
@@ -133,51 +145,93 @@ public abstract class CBuildConfiguration extends PlatformObject implements ICBu
+	 *
+	 * @param buildConfig Platform Build Configuration. Must not be null.
+	 * @param cBuildConfigName Name to give the CMakeBuildConfiguration. Must not be null.
+	 * @throws CoreException
+	 *   Toolchain is missing,
+	 *   Launch mode missing,
+	 *   Launch Target is missing
```
I have moved this to https://github.com/eclipse-cdt/cdt/issues/1077 and marked this conversation resolved.

jonahgraham reviewed (commented):
This is looking good. No real concerns, and all my comments are relatively minor.

> Design an api to get the ICBuildConfiguration name, generated in

I think you should defer this part to a future PR so we can progress this one to completion soon. It isn't ideal that the code is duplicated, but it was duplicated before you came along, so we don't have to fix it now.

betamaxbandit reviewed (commented):
still some work todo.

betamaxbandit commented:
DONE:
-  Complete ## API Changes in CDT 12.0. section of CHANGELOG-API.md
-  Design an api to get the ICBuildConfiguration name, generated in ICBuildConfigurationProvider.createBuildConfiguration, which is duplicated in the following classes, almost identically:
CMakeBuildConfigurationProvider
AutotoolsBuildConfigurationProvider
MakefileBuildConfigurationProvider
MesonBuildConfigurationProvider
- Complete CBuildConfigurationManagerTests
- Added ICBuildConfigurationProviderTests

TODO:
- To investigate:
In a Make project, changing the launch target to a new value causes this exception.
Does not happen when the project is CMake or Meson.

```
!ENTRY org.eclipse.cdt.build.gcc.core 4 0 2025-02-11 13:56:03.234
!MESSAGE C:\Users\a5107948\cdt-main20250110b\runtime-cdtruntime2\make\build\make.run.win32.x86_64.gen2\.sc15162304364447728639.cpp
!STACK 0
java.nio.file.NoSuchFileException: C:\Users\a5107948\cdt-main20250110b\runtime-cdtruntime2\make\build\make.run.win32.x86_64.gen2\.sc15162304364447728639.cpp
	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
	at java.base/sun.nio.fs.WindowsFileSystemProvider.newByteChannel(WindowsFileSystemProvider.java:234)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:379)
	at java.base/java.nio.file.Files.createFile(Files.java:657)
	at java.base/java.nio.file.TempFileHelper.create(TempFileHelper.java:136)
	at java.base/java.nio.file.TempFileHelper.createTempFile(TempFileHelper.java:159)
	at java.base/java.nio.file.Files.createTempFile(Files.java:878)
	at org.eclipse.cdt.build.gcc.core.GCCToolChain.getDefaultScannerInfo(GCCToolChain.java:459)
	at org.eclipse.cdt.core.build.CBuildConfiguration.getScannerInformation(CBuildConfiguration.java:787)
	at org.eclipse.cdt.internal.core.model.TranslationUnit.getScannerInfo(TranslationUnit.java:1070)
	at org.eclipse.cdt.internal.core.pdom.indexer.ProjectIndexerInputAdapter.getBuildConfiguration(ProjectIndexerInputAdapter.java:160)
	at org.eclipse.cdt.internal.core.pdom.AbstractIndexerTask.getScannerInfo(AbstractIndexerTask.java:1030)
	at org.eclipse.cdt.internal.core.pdom.AbstractIndexerTask.parseLinkage(AbstractIndexerTask.java:909)
	at org.eclipse.cdt.internal.core.pdom.AbstractIndexerTask.runTask(AbstractIndexerTask.java:572)
	at org.eclipse.cdt.internal.core.pdom.indexer.PDOMIndexerTask.run(PDOMIndexerTask.java:164)
	at org.eclipse.cdt.internal.core.pdom.indexer.PDOMRebuildTask.run(PDOMRebuildTask.java:94)
	at org.eclipse.cdt.internal.core.pdom.PDOMIndexerJob.run(PDOMIndexerJob.java:160)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

betamaxbandit commented:
> TODO:
> To investigate:
> In a Make project, changing the launch target to a new value causes this exception.

I investigated this and it is not to be related to my change.
I have created a bug for it in #1080

betamaxbandit commented:
@jonahgraham ,

I see the "Code Cleanliness Checks / build (pull_request) Failing after 3m" job failed for my PR, but in the detail I don't see what the problem is. Can you help please?

jonahgraham commented on core/org.eclipse.cdt.core.tests/model/org/eclipse/cdt/core/build/ICBuildConfigurationProviderTests.java:
```diff
@@ -0,0 +1,130 @@
+		 */
+		ICBuildConfiguration cBuildConfiguration = configManager.getBuildConfiguration(project, mockToolchain,
+				ILaunchManager.RUN_MODE, launchTargetManager.getLocalLaunchTarget(), new NullProgressMonitor());
+		assertThat(cBuildConfiguration, is(notNullValue()));
+
+		assertThat(cBuildConfiguration, instanceOf(CBuildConfiguration.class));
+		if (cBuildConfiguration instanceof CBuildConfiguration cbc) {
+			/*
+			 * Note, this extra test uses getName on CBuildConfiguration rather than the API ICBuildConfiguration
+			 */
+			String name = cbc.getName();
+			assertThat(name, is(expectedName));
+		}
+
+		assertThat(cBuildConfiguration, instanceOf(ICBuildConfiguration2.class));
+		if (cBuildConfiguration instanceof ICBuildConfiguration2 cbc2) {
+			/*
+			 * (2) Check last segment of build output directory name is expected.
+			 */
+			URI buildDirectoryURI = cbc2.getBuildDirectoryURI();
+			Path path = Paths.get(buildDirectoryURI);
+			String lastSegment = path.getFileName().toString();
+			assertThat(lastSegment, is(expectedName));
+		}
```
You can simplify this by removing the extra null and instanceof checks. In tests you can just let NPE and CCE happen without having to do the extra checks for it. 

```suggestion
		CBuildConfiguration cBuildConfig = (CBuildConfiguration) cBuildConfiguration;
		assertThat(cBuildConfig.getName(), is(expectedName));
		/*
		 * (2) Check last segment of build output directory name is expected.
		 */
		Path buildDirectory = cBuildConfig.getBuildDirectory();
		String lastSegment = buildDirectory.getFileName().toString();
		assertThat(lastSegment, is(expectedName));
```

betamaxbandit replied:
Great.
That's simpler and more concise then.
Also I like cBuildConfig.getBuildDirectory(); rather than cbc2.getBuildDirectoryURI();

jonahgraham commented on core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/ICBuildConfigurationManager.java:
```diff
@@ -49,54 +48,45 @@ boolean hasConfiguration(ICBuildConfigurationProvider provider, IProject project
+	 * @return new Platform Build configuration. Not null.
+	 * @throws CoreException Reasons include the reasons given in
+	 * {@link IProject#setDescription(org.eclipse.core.resources.IProjectDescription, IProgressMonitor)}
+	
```
Trailing white space here causes code cleanliness fail, but further than that lines in a javadoc should start with `*`. In this case as it is the last line I am suggesting deleting the whole line.
```suggestion
```

PS You can see this in the build output [here](https://github.com/eclipse-cdt/cdt/actions/runs/13264673819/job/37029075519#step:4:134) <-- is a link to the specific line, but if you click on Details link you should come to the same approximate place:

![image](https://github.com/user-attachments/assets/7a80fb0a-dbc1-469c-8443-2d09eb9a6a04)



This is a screenshot so you can see the key parts:

![image](https://github.com/user-attachments/assets/38a25896-df60-4043-a5e0-5c3e17b5556c)

betamaxbandit replied:
Ah got it. Thanks for the comprehensive help on that.
I'm missing formatting in comments 😢

jonahgraham reviewed (commented):
@betamaxbandit  There are a few small changes which I will apply manually. Please have a look and let me know if they are ok. And if so I will merge them

I did test autotools and it looks fine.

jonahgraham commented:
@betamaxbandit here is a [link](https://github.com/eclipse-cdt/cdt/pull/1076/commits/73ea9ae30bf48315a2fd99205ce6d88803404a2c) to to my application of the changes in the review tab. It can also be seen by clicking on the commit in the timeline above:

![image](https://github.com/user-attachments/assets/1780b132-3ece-43ae-829e-fb8c786a732b)

betamaxbandit reviewed (commented):
LGTM.
Hi @jonahgraham ,
thank you for all your suggestions. I agree with all of them.
I see "Apply review comments", but can't click it.
Please go ahead and commit these if that is what you meant.
Thanks John

betamaxbandit commented on core/org.eclipse.cdt.core.tests/model/org/eclipse/cdt/core/build/ICBuildConfigurationProviderTests.java:
```diff
@@ -0,0 +1,130 @@
+			/*
+			 * Note, this extra test uses getName on CBuildConfiguration rather than the API ICBuildConfiguration
+			 */
+			String name = cbc.getName();
```
Yes, that's it

betamaxbandit commented on core/org.eclipse.cdt.core.tests/model/org/eclipse/cdt/core/build/CBuildConfigurationManagerTests.java:
```diff
@@ -118,6 +137,73 @@ public void testResetCachedScannerInfoProvider() throws Exception {
+		 * Add a new Launch Target, which also sets it as the active launch target, thereby triggering the
+		 * CoreBuildLaunchBarTracker which ends up calling ICBuildConfiguration2.setActive() to set the active Build Configuration.
+		 */
+		ILaunchTarget launchTarget = launchTargetManager.addLaunchTarget(ILaunchTargetManager.localLaunchTargetTypeId,
```
Neat, I didn't think of that.

betamaxbandit commented on core/org.eclipse.cdt.core.tests/model/org/eclipse/cdt/core/build/CBuildConfigurationManagerTests.java:
```diff
@@ -118,6 +137,73 @@ public void testResetCachedScannerInfoProvider() throws Exception {
+
+		// Currently active launch target before we change anything is expected to be "Local"
+		assertThat(launchBarManager.getActiveLaunchTarget().getId(), is("Local"));
+		System.out.println("Active launch target id before=" + launchBarManager.getActiveLaunchTarget().getId());
```
Good spot.

betamaxbandit commented on NewAndNoteworthy/CHANGELOG-API.md:
```diff
@@ -10,6 +10,148 @@ Below is the detailed descriptions of API changes and mitigation efforts API con
 
 ## API Changes in CDT 12.0.
 
+### org.eclipse.cdt.core.build.ICBuildConfigurationProvider
```
Good idea.

betamaxbandit commented:
> @betamaxbandit There are a few small changes which I will apply manually. Please have a look and let me know if they are ok. And if so I will merge them
> 
> I did test autotools and it looks fine.

Great, thank you. I'm happy with all the changes you suggest.
Thanks for testing autotools.

jonahgraham commented:
@betamaxbandit I think this is ready to squash and merge. Let me know when you are done your final check.

betamaxbandit commented:
> @betamaxbandit I think this is ready to squash and merge. Let me know when you are done your final check.

Hi @jonahgraham ,
I've tested the latest commit and all looks fine. Please go ahead and merge.

jonahgraham commented:
Thanks @betamaxbandit for this significant contribution.
